pkg/operator/hive/hive.go

```diff
@@ -342,3 +345,83 @@ func (r *ReconcileHiveConfig) runningOnOpenShift(hLog log.FieldLogger) (bool, er
 
 	return len(list.APIResources) > 0, nil
 }
+
+func (r *ReconcileHiveConfig) cleanupLegacySyncSetInstances(hLog log.FieldLogger) error {
+	crdClient := r.dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1beta1",
+		Resource: "customresourcedefinitions",
+	})
+	syncSetInstanceCRD, err := crdClient.Get(context.Background(), "syncsetinstances.hive.openshift.io", metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		hLog.Debug("syncsetinstance crd has already been deleted")
+		return nil
+	case err != nil:
+		return errors.Wrap(err, "could not get the syncsetinstance CRD")
+	}
+	// Delete all the SyncSetInstance. List until there are no more SyncSetInstances to catch SyncSetInstances that may
```

Comment:
I have not tested this myself. I have been told that deleting the CRDs just leaves the CRs in etcd.

Reply:
A quick search shows that you are correct [1]. I will need to think about whether there is still value in doing the separate delete to cover the case where the syncsetinstance controller may still be running.

[1] https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#delete-a-customresourcedefinition

Reply:
Cool.

I would still expect you to do it in a loop-until-all-gone, which ought to cover the case of zombie ssi controller. We'll win eventually :)

Reply:
...and as with your current logic, once the CRD is thoroughly gone, the ssi controller will error or no-op any time it tries to do anything, which is fine.

Reply:
My concern is that a zombie controller may attempt to delete a resource from the target cluster during the delete of the CRD. Let's say that the hive operator (1) removes the finalizers from the CRs and then (2) removes the CRD. If the ssi controller is running, then between (1) and (2) the ssi controller could add the finalizer back to a CR. As part of (2), the CR will get the deletion timestamp, and the ssi controller will running the finalization logic on the CR, removing synced resources from the target cluster.

I will do some manual testing with CRDs to see what the behavior is. Right now, I am leaning towards leaving the logic as is. I would rather be overly cautious even if it is not as efficient (or even not necessary). I would hate to accidentally delete a resource from a customer's cluster.

It is probably better to add logic that waits for any older versions of hive-controllers to terminate before proceeding with this step. Then I would not worry about a zombie ssi controller. That was how I was going to do it originally, but since I had protections in place with deleting the CRs, it was not necessary.

Reply:
Looking back at the logic in the ssi controller, I see the path you're talking about. Agree that makes it important to do the steps in the order you have them (remove finalizers, then delete with the `resourceVersion` check). In the flow I suggested, the auto-deletion of the CR should be _immediate_ once its finalizers go away; but I sure wouldn't want to rely on it.

> It is probably better to add logic that waits for any older versions of hive-controllers to terminate before proceeding

Yeah. But probably not worth the development effort at this point to save, what, a few hundred `Delete()`s per hive, done one time?